clang/cursor.go

```diff
@@ -45,20 +46,54 @@ func (c Cursor) PlatformAvailability(availabilitySize int) (always_deprecated bo
  */
 type CursorVisitor func(cursor, parent Cursor) (status ChildVisitResult)
 
+type functionsMapping struct {
+	sync.Mutex
+
+	index int
+	funcs map[int]*CursorVisitor
+}
+
+func (fm *functionsMapping) register(f *CursorVisitor) int {
+	fm.Lock()
+	defer fm.Unlock()
+
+	fm.index++
+	for fm.funcs[fm.index] != nil {
+		fm.index++
+	}
+
+	fm.funcs[fm.index] = f
+
+	return fm.index
+}
+
+func (fm *functionsMapping) lookup(index int) *CursorVisitor {
+	fm.Lock()
+	defer fm.Unlock()
+
+	return fm.funcs[index]
+}
+
+func (fm *functionsMapping) unregister(index int) {
+	fm.Lock()
+	defer fm.Unlock()
+
+	delete(fm.funcs, index)
+}
+
+var visitors = &functionsMapping{
+	funcs: map[int]*CursorVisitor{},
+}
+
 // GoClangCursorVisitor calls the cursor visitor
 //export GoClangCursorVisitor
-func GoClangCursorVisitor(cursor C.CXCursor, parent C.CXCursor, cfct unsafe.Pointer) (status ChildVisitResult) {
-	fct := *(*CursorVisitor)(cfct)
+func GoClangCursorVisitor(cursor C.CXCursor, parent C.CXCursor, cfct *C.int) (status ChildVisitResult) {
```

Comment:
isn't this a bit of pointer aliasing? (and thus not safe against "strict aliasing rules" ?)
my `C/C++` is admittedly a bit rusty, but I think we could still keep the `cfct unsafe.Pointer` and cast it to a `*C.int`:

```go
fct := *(*C.int)(cfct)
```